data.c

```diff
@@ -525,11 +515,59 @@ backup_data_file(const char *from_root,
 
 		file->read_size += read_len;
 	}
+	/*
+	 * In incremental backup mode, put an additional data to remember
+	 * the last block number.
+	 * Without this, if a vacuum shrinks pages, which happens when one
+	 * or more pages at the end of a table are deleted, pg_rman has
+	 * no way to know those pages are shrinked and the deleted pages
+	 * comeback during pg_rman recovery.
+	 */
+	if (current.backup_mode == BACKUP_MODE_INCREMENTAL)
+	{
+		header.block = ++blknum;
+		header.endpoint = true;
+
+		/*
+		 * pg_rman processes backup data in page unit, so create a
+		 * dummy page
+		 */
+		memset(page.data + sizeof(header), 0, BLCKSZ - sizeof(header));
+#ifdef HAVE_LIBZ
+		if (compress)
+		{
+			doDeflate(&z, sizeof(header), sizeof(outbuf), &header, outbuf, in,
+					  out, &crc, &file->write_size, Z_NO_FLUSH);
+			doDeflate(&z, BLCKSZ, sizeof(outbuf), page.data, outbuf, in,
+					  out, &crc, &file->write_size, Z_NO_FLUSH);
+		}
+		else
+#endif
+		{
+		    if (fwrite(&header, 1, sizeof(header), out) != sizeof(header) ||
+				fwrite(page.data, 1, BLCKSZ - sizeof(header), out)
+					!= BLCKSZ - sizeof(header))
+			{
+				int errno_tmp = errno;
+				/* oops */
+				fclose(in);
+				fclose(out);
+				ereport(ERROR,
+					(errcode(ERROR_SYSTEM),
+					 errmsg("could not write at block %u of \"%s\": %s to put endpoint",
+						blknum, to_path, strerror(errno_tmp))));
+			}
+			PGRMAN_COMP_CRC32(crc, &header, sizeof(header));
+			PGRMAN_COMP_CRC32(crc, page.data, BLCKSZ - sizeof(header));
+
+			file->write_size += BLCKSZ;
+		}
+	}
 
 #ifdef HAVE_LIBZ
 	if (compress)
 	{
-		if (file->read_size > 0)
+		if (file->read_size > 0 || header.endpoint)
```

Comment:
Sorry for no heads up.

Without this condition, I faced a regression test error at RESTORE COMMAND TEST-0010 because of error in deflateEnd(). 

I image proper modifications on the way of deflating can make this condition unnecessary, but I haven't found the way.

Reply:
Ah, I've found the reason.

It's necessary when the page size was changed to 0 after a full backup.
(I verified it by TRUNCATE a table between full and incremental backup.)

In that case,  file->read_size remains 0, but this patch adds 0-padded page at the tail, so it needs to do doDeflate(&z, ...,  Z_FINISH) but skipped.

How about adding a comment like this one?

```
  * finalize zstream.
  *
  * NOTE: when added a 0-filled page, we didn't read any file but wrote
  * data, so need finalization.
```

Reply:
Ah, I see.  Thanks for explaining.  How about:

``
NOTE: We need to this even if we didn't read anything from the file but still had
to write the 0-filled dummy page.
``

Reply:
Thanks!
I'll modify it.

Reply:
Sorry, I meant: We need to "do" this...